File: src/main/java/info/skyblond/archivedag/util/SecurityUtils.java
package info.skyblond.archivedag.util;

import org.springframework.data.domain.Pageable;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;

import java.util.List;

public class SecurityUtils {
    public static final List<String> VALIDATE_ROLES = List.of(
            "ROLE_USER", "ROLE_VIEWER", "ROLE_UPLOADER", "ROLE_ADMIN"
    );

    public static Authentication getCurrentAuthentication() {
        return SecurityContextHolder.getContext().getAuthentication();
    }

    public static String getCurrentUsername() {
        return getCurrentAuthentication().getName();
    }

    public static boolean checkCurrentUserHasRole(String role) {
        Authentication auth = getCurrentAuthentication();
        if (auth == null) {
            return false;
        }
        return auth.getAuthorities().stream().anyMatch(a -> a.getAuthority().equals(role));
    }

    public static void requireValidateRole(String role) {
        if (!VALIDATE_ROLES.contains(role)) {
            throw new IllegalArgumentException("Sort properties not in range.");
        }
    }

    public static void requireSortPropertiesInRange(Pageable pageable, List<String> range) {
        if (!pageable.getSort().stream().allMatch(o -> range.contains(o.getProperty()))) {
            throw new IllegalArgumentException("Sort properties not in range.");
        }
    }
}

File: backend/src/main/kotlin/info/skyblond/archivedag/arudaz/utils/SecurityUtils.kt
package info.skyblond.archivedag.arudaz.utils

import org.springframework.data.domain.Pageable
import org.springframework.data.domain.Sort
import org.springframework.security.core.Authentication
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.context.SecurityContextHolder

val validatedRoles = listOf(
    "ROLE_USER", "ROLE_VIEWER", "ROLE_UPLOADER", "ROLE_ADMIN"
)

fun getCurrentAuthentication(): Authentication? {
    return SecurityContextHolder.getContext().authentication
}

fun getCurrentUsername(): String {
    return getCurrentAuthentication()!!.name
}

fun checkCurrentUserIsAdmin(): Boolean {
    val auth = getCurrentAuthentication() ?: return false
    return auth.authorities.stream().anyMatch { a: GrantedAuthority -> a.authority == "ROLE_ADMIN" }
}

fun requireValidateRole(role: String) {
    require(validatedRoles.contains(role)) { "Invalid role. Role can only be one of: $validatedRoles" }
}

fun requireSortPropertiesInRange(pageable: Pageable, range: List<String>) {
    require(pageable.sort.stream().allMatch { o: Sort.Order ->
        range.contains(
            o.property
        )
    }) { "Sort properties not in range." }
}
